### DISCUSSION TITLE
Question / Comment: Installing PyMuPDF with pip on alpine linux

### DISCUSSION THREAD

**Discussion by kharann** [pymupdf/PyMuPDF#1015] [Installation Topics]
Hi i keep trying to install PyMuPDF in a docker container running alpine linux.
I read around a bit and tried to add mupdf before pip installing without any success
```
apk add --virtual .build-deps gcc python3-dev musl-dev libffi-dev \
  # TODO workaround start
  && apk add libressl-dev \
  && apk add openssl-dev \
  && apk add python3-dev \
  && apk add mupdf \
  && pip install --upgrade pip  \
  && pip install PyMuPDF \
```
But it seems like im getting some weird kind of wheel error? Any guidance on the issue??
```
Collecting PyMuPDF
  Downloading https://files.pythonhosted.org/packages/27/9d/22195340c713be7e203af6551b5cf43b50f7394e5120e0f464485524ea02/PyMuPDF-1.16.6.tar.gz (165kB)
Building wheels for collected packages: PyMuPDF
  Building wheel for PyMuPDF (setup.py): started
  Building wheel for PyMuPDF (setup.py): finished with status 'error'
  ERROR: Command errored out with exit status 1:
   command: /usr/local/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-eimjk5e9/PyMuPDF/setup.py'"'"'; __file__='"'"'/tmp/pip-install-eimjk5e9/PyMuPDF/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-rmotb7uw --python-tag cp37
       cwd: /tmp/pip-install-eimjk5e9/PyMuPDF/
  Complete output (19 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.linux-x86_64-3.7
  creating build/lib.linux-x86_64-3.7/fitz
  copying fitz/__init__.py -> build/lib.linux-x86_64-3.7/fitz
  copying fitz/fitz.py -> build/lib.linux-x86_64-3.7/fitz
  copying fitz/utils.py -> build/lib.linux-x86_64-3.7/fitz
  running build_ext
  building 'fitz._fitz' extension
  creating build/temp.linux-x86_64-3.7
  creating build/temp.linux-x86_64-3.7/fitz
  gcc -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -DTHREAD_STACK_SIZE=0x100000 -fPIC -I/usr/include/mupdf -I/usr/local/include/mupdf -I/usr/local/include/python3.7m -c fitz/fitz_wrap.c -o build/temp.linux-x86_64-3.7/fitz/fitz_wrap.o
  fitz/fitz_wrap.c:2732:10: fatal error: fitz.h: No such file or directory
   #include <fitz.h>
            ^~~~~~~~
  compilation terminated.
  error: command 'gcc' failed with exit status 1
  ----------------------------------------
  ERROR: Failed building wheel for PyMuPDF
  Running setup.py clean for PyMuPDF
Failed to build PyMuPDF
Installing collected packages: PyMuPDF
    Running setup.py install for PyMuPDF: started
    Running setup.py install for PyMuPDF: finished with status 'error'
    ERROR: Command errored out with exit status 1:
     command: /usr/local/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-eimjk5e9/PyMuPDF/setup.py'"'"'; __file__='"'"'/tmp/pip-install-eimjk5e9/PyMuPDF/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-be3wrjpu/install-record.txt --single-version-externally-managed --compile
         cwd: /tmp/pip-install-eimjk5e9/PyMuPDF/
    Complete output (19 lines):
    running install
    running build
    running build_py
    creating build
    creating build/lib.linux-x86_64-3.7
    creating build/lib.linux-x86_64-3.7/fitz
    copying fitz/__init__.py -> build/lib.linux-x86_64-3.7/fitz
    copying fitz/fitz.py -> build/lib.linux-x86_64-3.7/fitz
    copying fitz/utils.py -> build/lib.linux-x86_64-3.7/fitz
    running build_ext
    building 'fitz._fitz' extension
    creating build/temp.linux-x86_64-3.7
    creating build/temp.linux-x86_64-3.7/fitz
    gcc -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -DTHREAD_STACK_SIZE=0x100000 -fPIC -I/usr/include/mupdf -I/usr/local/include/mupdf -I/usr/local/include/python3.7m -c fitz/fitz_wrap.c -o build/temp.linux-x86_64-3.7/fitz/fitz_wrap.o
    fitz/fitz_wrap.c:2732:10: fatal error: fitz.h: No such file or directory
     #include <fitz.h>
              ^~~~~~~~
    compilation terminated.
    error: command 'gcc' failed with exit status 1
    ----------------------------------------
ERROR: Command errored out with exit status 1: /usr/local/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-eimjk5e9/PyMuPDF/setup.py'"'"'; __file__='"'"'/tmp/pip-install-eimjk5e9/PyMuPDF/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-be3wrjpu/install-record.txt --single-version-externally-managed --compile Check the logs for full command output.
```

**Comment by JorjMcKie** [comment]
I have no experience with docker unfortunately. All I can do is recommend you roughly follow the PyMuPDF wheel creation process on Linux machines of Travis CI.
Here is a link: https://travis-ci.org/JorjMcKie/py-mupdf/jobs/606419367.

You error above shows that the used ``setup.py`` does not reference the include and lib folders of the MuPDF installation correctly.
So I **_think_** that a docker container must include both, the download and installation / generation of **MuPDF**, and then the ``python setup.py install`` of **PyMuPDF** referencing that installation.

**Comment by cges30901** [comment]
I am not familiar with Alpine Linux, but `fitz.h` seems to belong to `mupdf-dev`. Did you install this package?

**Comment by JorjMcKie** [comment]
>  am not familiar with Alpine Linux, but fitz.h seems to belong to mupdf-dev.

@cges30901 - thanks for stepping in here.
Yes, ``fitz.h`` is part of the MuPDF sources. MuPDF must be installed before PyMuPDF can be generated with the MuPDF sources and generated libs being available ...

**Comment by kharann** [comment]
> I am not familiar with Alpine Linux, but `fitz.h` seems to belong to `mupdf-dev`. Did you install this package? 
yeh i installed mupdf-dev too

**Comment by JorjMcKie** [comment]
@kharann - I think you should dilligently verify ``setup.py``:

* ``include_dirs`` must contain the installed MuPDF locations ``/mupdf/include`` and ``/mupdf/include/mupdf``. If this is done correctly, ``fitz.h`` will be found.
* ``library_dirs`` must contain the folder where the ``*.o`` files of the MuPDF generation landed.

**Comment by cges30901** [comment]
How did you install mupdf (from Alpine Linux repository or you built from source)? What version of mupdf and Alpine Linux are you using?

**Comment by JorjMcKie** [comment]
@kharann  - how are you doing ... any progress / success?

**Comment by kharann** [comment]
I decided to switch the docker image to Debian instead of using Alpine. I tried a few different things with little success on the alpine image. Thank you for the help nevertheless! It seems to work well with the Debian image tho!

**Comment by JorjMcKie** [comment]
Thank your for the feedback.
Assuming your okay, will close this now.

**Comment by cges30901** [comment]
Alpine Linux uses musl libc, which is not compatible with `manylinux`. So it is more difficult to install PyMuPDF on Alpine Linux.

**Comment by harryadel** [comment]
Would you kindly share the debian Dockerfile @kharann?

**Comment by harryadel** [comment]
Quick update after fiddling around. I got it working by adding `pip3 install --upgrade pip` to the docker file!

Here's the full file for reference:
```
FROM ubuntu:18.04
  
ENV DEBIAN_FRONTEND=noninteractive

RUN apt-get -y update && \
 apt-get -y upgrade && \
 apt-get -y dist-upgrade && \
 apt-get -y autoremove

# 2. Install Python3 and required libraries

RUN apt-get install -y python-dev python-tk python-numpy python3-dev python3-tk python3-numpy libpython3-all-dev python3-pip libpython3-all-dev python3-all && \
 apt-get install -y libzbar0 && \
 apt install -y poppler-utils && pip3 install --upgrade pip

RUN apt-get install -y locales && locale-gen en_US.UTF-8
ENV LANG='en_US.UTF-8' LANGUAGE='en_US:en' LC_ALL='en_US.UTF-8'

WORKDIR /app

# We copy just the requirements.txt first to leverage Docker cache
COPY requirements.txt .

RUN pip3 install -r requirements.txt

COPY . /app

ENTRYPOINT [ "python3" ]

CMD [ "app.py" ]

```

**Comment by rocadev** [comment]
Check out this repo: https://github.com/jay-eff/mutool
There is a way for building MuPDF command line tool "mutool" based on Alpine:3

**Comment by rocadev** [comment]
there are already mupdf apk for alpine
https://pkgs.alpinelinux.org/package/edge/community/x86_64/mupdf
https://pkgs.alpinelinux.org/package/edge/community/x86/mupdf-tools

**Comment by Franr** [comment]
We end up adapting the `ubuntu` installation script that is already on the repo:

```bash
#!/bin/bash
set -euo pipefail

wget https://github.com/pymupdf/PyMuPDF/archive/1.18.6.tar.gz
tar -xzf 1.18.6.tar.gz && cd PyMuPDF-1.18.6

wget https://mupdf.com/downloads/archive/mupdf-1.18.0-source.tar.gz
tar -zxvf mupdf-1.18.0-source.tar.gz
cd mupdf-1.18.0-source
# replace files in mupdf source
cp ../fitz/_config.h include/mupdf/fitz/config.h

export CFLAGS="-fPIC"
make HAVE_X11=no HAVE_GLFW=no HAVE_GLUT=no prefix=/usr/local
make HAVE_X11=no HAVE_GLFW=no HAVE_GLUT=no prefix=/usr/local install

cd ..

python setup.py build
python setup.py install
```

**Comment by LuisVillaC** [comment]
Hi, this work for me, and only work with python:3.8.10 or greater and with Pymupdf:1.18.14:

**1). Only if you don't have python installed:**
`apk add --update py-pip`

**2). Install dependencies:**
   ```
apk add --no-cache \
   python3-dev \
   mupdf-dev \
   gcc \
   libc-dev \
   musl-dev \
   jbig2dec \
   openjpeg-dev \
   jpeg-dev \
   harfbuzz-dev
```
**3). Create Shortcut:**
  `ln -s /usr/lib/libjbig2dec.so.0 /usr/lib/libjbig2dec.so`

**4). Install PymuPdf:**
    `pip install pymupdf`

**Only if you want to do everything at the same time**:

```
apk add --update py-pip \
&& apk add --no-cache \
python3-dev \
mupdf-dev \
gcc \
libc-dev \
musl-dev \
jbig2dec \
openjpeg-dev \
jpeg-dev \
harfbuzz-dev \
&& ln -s /usr/lib/libjbig2dec.so.0 /usr/lib/libjbig2dec.so \
&& pip install pymupdf
```

**Comment by MariaFjodorowa** [comment]
The abovementioned setup has almost worked for me, but I had to add make and g++ (and have also added swig as said in the [docs](https://pymupdf.readthedocs.io/en/latest/installation.html#installation-when-a-suitable-wheel-is-not-available) ) .

The whole working setup looks like:

```
FROM python:3.11-alpine
RUN apk update && apk add --no-cache \
mupdf-dev \
gcc \
libc-dev \
musl-dev \
jbig2dec \
openjpeg-dev \
jpeg-dev \
harfbuzz-dev \
swig \
make \
g++ &&  ln -s /usr/lib/libjbig2dec.so.0 /usr/lib/libjbig2dec.so && pip install --no-cache-dir pymupdf==1.21.1
```

**Reply by fgplastina** [reply]
It install ok but it said fitz can't be imported. I don't standard why

**Reply by julian-smith-artifex-com** [reply]
Please give the command that you ran to install PyMuPDF and the complete output. And also include the Python code that you're testing it with, with the complete output.

**Reply by fgplastina** [reply]
Now i got it working, i was using venv inside my docker and was giving some troubles with this. Removing that and running @MariaFjodorowa code works just fine. Thank you Julian.

**Reply by ZhimaoLin** [reply]
We use `3.11.5-alpine` base image, and we only need to install the following packages for `pymupdf==1.21.1`.

```
RUN apk update && apk add --no-cache \
gcc \
musl-dev \
openjpeg-dev \
jpeg-dev \
swig \
make \
g++
```

Also we did NOT run `ln -s /usr/lib/libjbig2dec.so.0 /usr/lib/libjbig2dec.so` since the file `libjbig2dec.so` already existed. 
We use `pipenv` to install `pymupdf`

**Comment by brunofuentes** [comment]
anybody was able to run on 3.12-alpine? I'm trying to use pymupdf==1.24.11, tried all the above but haven't got it to work so far :(
